docs/deploying.rst

```diff
@@ -25,9 +25,9 @@ Here's an example of what that ``asgi.py`` might look like:
     from django.conf.urls import url
     from django.core.asgi import get_asgi_application
 
-    # Fetch Django ASGI application early to ensure AppRegistry is populated
-    # before importing consumers and AuthMiddlewareStack that may import ORM
-    # models. 
+    # Fetch Django ASGI application early to ensure Django settings are 
+    # configured and the AppRegistry is populated before importing consumers
+    # and AuthMiddlewareStack that may import ORM models or settings.
```

Comment:
importing settings doesnt cause any problem, this is misleading

Reply:
I have tested import of settings with this code and it works without any problem: 

```import os
from django.conf import settings

# randomly print some settings
print(settings, settings.STATIC_URL)


from django.core.asgi import get_asgi_application
os.environ.setdefault("DJANGO_SETTINGS_MODULE", "myapp.settings")

django_asgi_app = get_asgi_application()


from channels.auth import AuthMiddlewareStack
from channels.routing import ProtocolTypeRouter, URLRouter
from myapp.chat.routing import websocket_urlpatterns

application = ProtocolTypeRouter(
    {
        "http": django_asgi_app,
        "websocket": AuthMiddlewareStack(URLRouter(websocket_urlpatterns)),
    }
)
```
output
```
$ gunicorn myapp.asgi:application -k uvicorn.workers.UvicornWorker -b 127.0.0.1:8000
[2021-04-21 20:39:28 +0700] [5328] [INFO] Starting gunicorn 20.1.0
[2021-04-21 20:39:28 +0700] [5328] [INFO] Listening at: http://127.0.0.1:8000 (5328)
[2021-04-21 20:39:28 +0700] [5328] [INFO] Using worker: uvicorn.workers.UvicornWorker
[2021-04-21 20:39:28 +0700] [5370] [INFO] Booting worker with pid: 5370
<Settings "myapp.settings.dev"> /static/
```

Reply:
Its not misleading at all. If you decide to lose the variable name "django_asgi_app", and just call the function like this:

import X from projectsettings
import module.that.accesses.project.settings

application = ProtocolTypeRouter({
    # Django's ASGI application to handle traditional HTTP requests
    "http": get_asgi_application(),
})

then it will throw irrational errors, because you are trying to import a Django setting, without Django being initialised in this codespace

you MUST execute the function at the top of the page, and store it in a variable

Reply:
Perhaps `use` rather than `import` in the last line.

Reply:
I didn't test with any settings directly in the asgi.py file, I had included a module that didn't access the ORM models, but did access project.settings

As soon as you import a module that uses project.settings in it (eg, imagine you grab something from the settings to use in your consumer, and you import the consumer to include in the URL route) it breaks

Reply:
(The point really is the `django.setup()` is called, but I don't want to go into that in the Channels docs — this should be enough to help beginners, but folks should know their Django...)

Reply:
It's really difficult to reason about anything if you provide only pseudocode like only thing only pseudocode like `import module.that.accesses.project.settings` instead of the traceback you are getting.

If your code is really not accessing any models, then that should work. If it doesn't, then it's a bug in Django initialization process, and you should open it as a bug in Django bugtracker with a minimal reproducible example, instead of merely rephrasing instructions in one of Django apps.

Reply:
In your example of "randomly print some settings" will work, because you are assigning the `get_asgi_application()` to a variable, and passing that to the protocol router.

Try the exact same code snippet, but with the variable removed, and executing `get_asgi_application()` directly in the router.

Without a specific pointer, people coming to this project may not know that you MUST execute it before passing it to the router.
I'm just trying to help people in the future.

If you decide a better wording that my replacement, cool, I just want others to not have the same migraine I have right now.